src/ffi/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,147 @@
+/// to call this function when finished with a quaternion
+#[no_mangle]
+pub unsafe extern "C" fn free_quaternion_memory(ptr: *mut Quaternion) {
+    if ptr.is_null() {
+        return;
+    }
+    Box::from_raw(ptr);
+}
+
+/// Get the components of a quaternion as a list of C doubles, the order of the
+/// components will be (real, i, j, k)
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_get_components(quat_ptr: *const Quaternion) -> *const c_double {
+    null_pointer_check!(quat_ptr);
+    let quat = *quat_ptr;
+    let components: [c_double;4] = [quat.real, quat.i, quat.j, quat.k];
+    Box::into_raw(Box::new(components)) as *const _
+}
+
+/// Set the real component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_real(quat_ptr: *mut Quaternion, real: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.real = real
+}
+
+/// Set the i component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_i(quat_ptr: *mut Quaternion, i: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.i = i
+}
+
+/// Set the j component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_j(quat_ptr: *mut Quaternion, j: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.j = j
+}
+
+/// Set the k component of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_k(quat_ptr: *mut Quaternion, k: f64) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.k = k;
+}
+
+/// Set all of the components of an existing quaternion stored at the address
+/// of a pointer
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_components(
+    quat_ptr: *mut Quaternion, real: f64, i: f64, j: f64, k: f64
+) {
+    null_pointer_check!(quat_ptr);
+    let quat = &mut*quat_ptr;
+    quat.real = real;
+    quat.i = i;
+    quat.j = j;
+    quat.k = k;
+}
+
+/// Set the imaginary components of an existing quaternion stored at
+/// the address of a pointer (quat_ptr) from the components of a 3-vector
+/// (stored at vec_ptr). The convention is x -> i, y -> j, z -> k
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_set_imag_from_vector(quat_ptr: *mut Quaternion, vec_ptr: *const Vector3) {
+    null_pointer_check!(quat_ptr);
+    null_pointer_check!(vec_ptr);
+    let mut quat = *quat_ptr;
+    let imag = *vec_ptr;
+    quat.set_imag_from_vector(imag);
+}
+
+/// Copies the imaginary components to a 3-vector (using x -> i, y -> j
+/// z -> k) and returns a pointer to the memory address of the resulting
+/// vector
+#[no_mangle]
+pub unsafe extern "C" fn quaternion_get_imaginary_vector(quat_ptr: *const Quaternion) -> *mut Vector3 {
+    null_pointer_check!(quat_ptr);
+    let quat = *quat_ptr;
```

Comment:
here and in some other places you are copying the quaternion (let quat = *quat_ptr) and then getting the image is it on purpose?

Reply:
it's not, it's a mistake